hetmech/degree_weight.py

```diff
@@ -1,187 +1,642 @@
+            [0, 3], [3, 4], [4, 5]]
+
+    elif category in ('disjoint', 'short_repeat', 'long_repeat'):
+        indices = sorted([[metanodes.index(i), len(metapath) - list(
+            reversed(metanodes)).index(i)] for i in repeated])
+        indices = add_head_tail(metapath, indices)
+        # handle middle cases with non-repeated nodes between disjoint regions
+        # Eg. [[0,2], [3,4]] -> [[0,2],[2,3],[3,4]]
+        inds = []
+        for i, v in enumerate(indices[:-1]):
+            inds.append(v)
+            if v[-1] != indices[i + 1][0]:
+                inds.append([v[-1], indices[i + 1][0]])
+        indices = inds + [indices[-1]]
+
+    elif category in ('BAAB', 'BABA', 'other', 'interior_complete_group'):
+        nodes = set(metanodes)
+        repeat_indices = (
+            [[i for i, v in enumerate(metanodes)
+              if v == metanode] for metanode in nodes])
+        repeat_indices = [i for i in repeat_indices if len(i) > 1]
+        simple_repeats = [i for group in repeat_indices for i in group]
+        inds = []
+        for i in repeat_indices:
+            if len(i) == 2:
+                inds += i
+            if len(i) > 2:
+                inds.append(i[0])
+                inds.append(i[-1])
+                for j in i[1:-1]:
+                    if (j - 1 in simple_repeats and j + 1 in simple_repeats) \
+                            and not (j - 1 in i and j + 1 in i):
+                        inds.append(j)
+        inds = sorted(inds)
+        seconds = inds[1:] + [inds[-1]]
+        indices = list(zip(inds, seconds))
+        indices = [i for i in indices if len(set(i)) == 2]
+        indices = add_head_tail(metapath, indices)
+
+    segments = [metapath[i[0]:i[1]] for i in indices]
+    segments = [i for i in segments if i]
     segments = [metagraph.get_metapath(metaedges) for metaedges in segments]
-    duplicates = [range_to_duplicate.get(range_, None) for range_ in ranges]
-    return segments, duplicates
+    # eg: B CC ABA
+    if category == 'interior_complete_group':
+        segs = []
+        for i, v in enumerate(segments[:-1]):
+            if segments[i + 1].source() == segments[i + 1].target():
+                edges = v.edges + segments[i + 1].edges + segments[i + 2].edges
+                segs.append(metagraph.get_metapath(edges))
+            elif v.source() == v.target():
+                pass
+            elif segments[i - 1].source() == segments[i - 1].target():
+                pass
+            else:
+                segs.append(v)
+        segs.append(segments[-1])
+        segments = segs
+    return segments
+
+
+def _remove_diag(mat):
+    """Set the main diagonal of a square matrix to zeros."""
+    mattype = type(mat)
+    if mattype == numpy.ndarray:
+        mattype = numpy.array
+    assert mat.shape[0] == mat.shape[1]  # must be square
+    return mat - mattype(numpy.diag(mat.diagonal()), dtype=numpy.float64)
```

Comment:
Consider [this method](https://github.com/greenelab/hetmech/blob/775ef9b38f549dfd147104c66a256ae9c7df4605/hetmech/degree_weight.py#L155-L157)